File: 53-maximum-subarray/53-maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
    if(nums.length === 0) return 0
    
    let curr = 0, max = nums[0];
    
    for(let i = 0; i < nums.length; i++){
        let num = nums[i]
        curr = Math.max(num, curr + num)
        max = Math.max(max, curr)
    }
    return max
};

File: 53-maximum-subarray/53-maximum-subarray.rb
# @param {Integer[]} nums
# @return {Integer}
def max_sub_array(nums)
    curr, max = 0, nums[0]
    
    nums.each do |num|
        curr = [curr+num,num].max
        max = [max, curr].max
    end
    max
end